security/vault.go

```diff
@@ -0,0 +1,242 @@
+package security
+
+import (
+	"bytes"
+	"crypto/aes"
+	"crypto/cipher"
+	"crypto/rand"
+	"encoding/base64"
+	"errors"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"sync"
+
+	"github.com/gaia-pipeline/gaia"
+)
+
+const (
+	vaultName = ".gaia_vault"
+)
+
+// Vault is a secret storage for data that gaia needs to store encrypted.
+type Vault struct {
+	Path string
+	Cert []byte
+	data map[string][]byte
+	sync.RWMutex
+}
+
+// NewVault creates a vault which is a simple k/v storage medium with AES encryption.
+// The format is:
+// KEY=VALUE
+// KEY2=VALUE2
+func NewVault() (*Vault, error) {
+	v := new(Vault)
+	// Creating vault file
+	vaultPath := filepath.Join(gaia.Cfg.VaultPath, vaultName)
+	if _, osErr := os.Stat(vaultPath); os.IsNotExist(osErr) {
+		gaia.Cfg.Logger.Info("vault file doesn't exist. creating...")
+		_, err := os.Create(vaultPath)
+		if err != nil {
+			gaia.Cfg.Logger.Error("failed creating vault file: ", err.Error())
+			return nil, err
+		}
+	}
+
+	// Setting up certificate key content
+	c, err := InitCA()
```

Comment:
Don't create the CA here. It will be already created in the main.go file. You should pass this instance via Parameter to `NewVault`.

Reply:
Yeah, but I'm always using NewVault, and Ca won't be something global, and it shouldn't be. A call to init, should just be idempotent, OR... be a singleton which still is a global 🤔 Not sure.